File: 46_Permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        total_set = []
        def helper(current_nums, current_set=[]):
            if not current_nums:
                total_set.append(current_set)
            
            for i in range(len(current_nums)):
                helper(current_nums[0:i]+current_nums[i+1:], 
                current_set+[current_nums[i]])
            
            return total_set
        helper(nums, [])
        return total_set

File: cpp/46_Permutations.cpp
class Solution {
public:
    vector<vector<int>> results;
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> curr;
        unordered_set<int> nums_set;
        for (int n : nums) nums_set.insert(n);
        helper(nums, nums_set, curr);
        return results;
    }
    
    void helper(vector<int>& nums, unordered_set<int>& nums_set, vector<int>& curr) {
        if (nums_set.empty()) {
            results.push_back(vector<int>(curr));
            return;
        }
        
        for (int n : nums) {
            if (nums_set.find(n) != nums_set.end()) {
                nums_set.erase(n);
                curr.push_back(n);
                helper(nums, nums_set, curr);
                nums_set.insert(n);
                curr.pop_back();
            }
        }
    }
};
